lib/spack/spack/cmd/__init__.py

```diff
@@ -277,27 +277,52 @@ def disambiguate_spec_from_hashes(spec, hashes, local=False, installed=True, fir
     else:
         matching_specs = spack.store.STORE.db.query(spec, hashes=hashes, installed=installed)
     if not matching_specs:
-        tty.die("Spec '%s' matches no installed packages." % spec)
+        tty.die(f"Spec '{spec}' matches no installed packages.")
 
     elif first:
         return matching_specs[0]
 
-    ensure_single_spec_or_die(spec, matching_specs)
+    return get_single_spec_or_maybe_die(spec, matching_specs)
 
-    return matching_specs[0]
 
+def get_single_spec_or_maybe_die(
+    spec: spack.spec.Spec, matching_specs: List[spack.spec.Spec]
+) -> spack.spec.Spec:
+    if len(matching_specs) == 1:
+        return matching_specs[0]
+
+    if len(matching_specs) == 0:
+        tty.die(f"Spec '{spec}' matches no packages.")
+
+    if sys.stdin.isatty():
+        fmt = "{name}{@version}{%compiler.name}{@compiler.version}{arch=architecture}{/hash:7}"
+    else:
+        fmt = "{hash:7} {name}{@version}{%compiler.name}{@compiler.version}{arch=architecture}"
+    args = (
+        colorize("@*b{" f"[{i+1}]" "} ") + s.cformat(fmt) for (i, s) in enumerate(matching_specs)
+    )
 
-def ensure_single_spec_or_die(spec, matching_specs):
-    if len(matching_specs) <= 1:
-        return
+    if not sys.stdin.isatty():
```

Comment:
This isn't sufficient for deciding whether to be interactive or not.  There are terminals (like GHA, GitLab, etc.) for which `isatty()` is true but which shouldn't just hang if there isn't a human on the other end.  So the interactive part here should really be behind ` --interactive` or `-i` flag, and `spack load` should continue to fail fast by default if it doesn't find a match -- otherwise peoples' scripts are going to hang unexpectedly instead of failing.